app/controllers/transfer_packages_controller.rb

```diff
@@ -1,4 +1,59 @@
 class TransferPackagesController < ApplicationController
+  include Concerns::ViewHelper
+
+  helper_method :samples_data
+
+  def new
+    @view_helper = view_helper({ save_back_path: true })
+    @can_update = true
+
+    @transfer_package = TransferPackage.new(sender_institution: @navigation_context.institution)
+    @available_institutions = available_institutions
+  end
+
+  def create
+    params = transfer_package_params
+    params.merge!({
+      sender_institution: @navigation_context.institution,
+    })
+    if id = params.delete(:receiver_institution).presence
+      params[:receiver_institution] = Institution.find(id)
+    end
```

Comment:
The whole method is _very_ complex, and mostly because of a heavy manipulation of params. You could improve readability by relying more on rails' conventions, and using records.

For example:

- use a `receiver_institution_id` param name;
- instantiate the package through `@navigation_context.institution` (declare the association if it doesn't already);
- build the package object immediately, add sample transfers to it, and eventually save it all;
- rely on validations to do data integrity checks (e.g. sample transfer can't have a QC sample);
- we don't need attributes for sample transfers but the sample UUID, so could the frontend send a mere `params[:sample_uuids]` array?

Then the method would be more straightforward, pushing data details into the models:

```ruby
@package_transfer = @navigation_context
  .institution
  .package_transfers
  .build(package_transfer_params)

samples = Sample
  .within(@navigation_context.institution)
  .find_all_by_any_uuid(params[:sample_uuids])

samples.each do |sample|
  sample_transfer = @package_transfer.sample_transfers.build(sample: sample)

  unless authorize_resource?(sample, UPDATE_SAMPLE)
    sample_transfer.errors.add :sample, :unauthorized
  end
end

# NOTE: could push this validation to PackageTransfer
if @package_transfer.sample_transfers.empty?
  @package_transfer.errors.add :sample_transfers, :empty
elsif @package_transfer.save
  redirect_to # ...
  return
end

# ...
render "new"
```

Reply:
Good points. I'm not happy about it either. Ended up with this as a reasonably find solution after already trying some alternatives. I'll give it another go and see what can be improved.

> use a `receiver_institution_id param` name;

IIRC this didn't seem to work well with the CdxSelect field. 

> * build the package object immediately, add sample transfers to it, and eventually save it all;
> * we don't need attributes for sample transfers but the sample UUID, so could the frontend send a mere params[:sample_uuids] array?

The tradeoff for that is to do some additional things for the sample transfers (such as error handling). Passing in the sample transfers as nested attributes is actually a nice solution. It would be great if it didn't require some extra handling. Maybe I can get rid of that.

Reply:
I achieved improvements on most of the mentioned points.

* I don't think it makes much sense to introduce an association just for creating a transfer package from it. Setting `sender_institution` explicitly is completely fine. I changed it to not go through params, though.
* I'm keeping the nested attributes for sample transfers because it's a neater solution IMO. It easily allows editing and removing transfers if that should ever be necessary.
* Changed sample reference to use the (internal) ID. There's no need for UUID here. It's just more overhead.